### ISSUE TITLE
Need help using global appModel/understanding data binding

### ISSUE THREAD

**Issue by sangupta** [developit/preact#901]
Hi - I am starting to learn preact/vdom/data-binding internals, and as an experiment looking to port automatic data-binding (like in `vue.js` framework by the use of watchers) to `preact`.  It would be on the lines of generating a watch expression for all dependencies in a component and then subscribing to it by using `Object proxies` whenever an attribute changes. The idea is to implement something like this:

A global object representing the state of various pieces:
```json
var appModel = {
  'user': {  },
  'tweets': [ ],
  'notifications': [],
  'followers': []
}
```

And, different components watching/binded over parts of the above model:

```javascript
Preact.render(<ProfileBox user={ appModel.user } />, document.getElementById('profile'));
Preact.render(<TweetList list={ appModel.tweets } />, document.getElementById('tweetList')); 
```

1. What would be a good starting point to explore this? 
2. Are there ways already with `preact` to achieve something like above? 
3. Are there any major challenges that are foreseen in developing an example as above.

**Comment by yaodingyd** [comment]
Vue is build with mutation in mind by converting data model's to getter/setter for dependency tracking and change notifications. While preact follows react's immutability and state cannot be changed. Using a Object proxy is kind of defeating the purpose here.

However, you can have one object converted to getter/setter, with setter doing `setState`, and getter getting the state.

**Comment by sangupta** [comment]
Hi @yaodingyd - thanks for the info. As I mentioned I would like to add binding via mutation to preact if needed, as an experiment to see how a single global object model can be used for state changes.

I will try the approach mentioned above by converting setters to `setState` and see if that works as expected.

**Comment by yaodingyd** [comment]
If you are looking for a single global object model for state change, then it's exactly what Redux is doing: one single data store which is the sole source of truth. But I think the focus of what you want to achieve is notification through mutation, not to call `setState` repeatedly.

**Comment by sangupta** [comment]
Am looking at both. So I wanted to try to convert to use notifications for object mutation handling than calling `setState` repeatedly. To build a working example I chose to have a global object model. I haven't used Redux before - let me take a look on how it works. Thanks a lot for the guidance.

**Comment by developit** [comment]
MobX is also a decent source of inspiration here. If you're looking at setting up a subscription mechanism (eg: via `context`) rather than using Preact's own state management, you should be able to use `Component#forceUpdate()` or `Component#setState({})` to trigger arbitrary re-renders of a wrapper component in response to your state object being mutated.

@sangupta you might find this Gist I created a while ago interesting:  it's more like Redux than Vue's state management, but shows you how to set up subscriptions:
https://gist.github.com/developit/55c48d294abab13a146eac236bae3219

Also, if you're just looking for mutations that automatically trigger rendering, you might find this tiny implementation of Glimmer's `@tracked` decorator interesting - it lets you create observed properties that are treated as Preact state:
https://gist.github.com/developit/006c58bba02be99e9a2ea1aca536dd33

**Comment by sangupta** [comment]
@developit thanks a bunch for the advice. I had worked the `@tracked` decorator approach after suggestions from @yaodingyd. 

I assume the approach mentioned in the first gist is the one that is being used in https://github.com/nightwolfz/mobx-preact - it uses the same provider concept. 

Anyhow, it has been a great learning curve for me with `preact` till now. Thanks @developit for the wonderful framework.

@developit  I have added some notes (JSDoc) to functions in `preact` code to make it more meaningful for newbee's like me - should I raise a PR for them?
